alacritty/src/renderer/rects.rs

```diff
@@ -190,3 +195,208 @@ impl RenderLines {
         }
     }
 }
+
+/// Shader sources for rect rendering program.
+pub static RECT_SHADER_F_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/res/rect.f.glsl");
+pub static RECT_SHADER_V_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/res/rect.v.glsl");
+static RECT_SHADER_F: &str = include_str!("../../res/rect.f.glsl");
+static RECT_SHADER_V: &str = include_str!("../../res/rect.v.glsl");
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Rgba {
+    r: u8,
+    g: u8,
+    b: u8,
+    a: u8,
+}
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy)]
+struct Vertex {
+    // Normalized screen coordinates.
+    x: f32,
+    y: f32,
+
+    // Color.
+    color: Rgba,
+}
+
+#[derive(Debug)]
+pub struct RectRenderer {
+    // GL buffer objects.
+    vao: GLuint,
+    vbo: GLuint,
+
+    program: RectShaderProgram,
+
+    vertices: Vec<Vertex>,
+}
+
+impl RectRenderer {
+    /// Update program when doing live-shader-reload.
+    pub fn set_program(&mut self, program: RectShaderProgram) {
+        self.program = program;
+    }
+
+    pub fn new() -> Result<Self, renderer::Error> {
+        let mut vao: GLuint = 0;
+        let mut vbo: GLuint = 0;
+        let program = RectShaderProgram::new()?;
+
+        unsafe {
+            // Allocate buffers.
+            gl::GenVertexArrays(1, &mut vao);
+            gl::GenBuffers(1, &mut vbo);
+
+            gl::BindVertexArray(vao);
+
+            // VBO binding is not part of VAO itself, but VBO binding is stored in attributes.
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+
+            let mut attribute_offset = 0;
+
+            // Position.
+            gl::VertexAttribPointer(
+                0,                            // Attribute location.
+                2,                            // Component count: 2. x and y.
+                gl::FLOAT,                    // Attribute type: f32
+                gl::FALSE,                    // Not normalized.
+                size_of::<Vertex>() as i32,   // Stride.
+                attribute_offset as *const _, // Offset of Vertex::x is zero.
```

Comment:
I find the inline documentation here a bit strange. What is the purpose of this?

The `glVertexAttribPointer` function should have the different inputs documented already, correct? So if we need to document anything, it would only be things that are relevant to our individual usecase. Like x/y, r/g/b/a for example.

Reply:
I also found them a bit strange, though, it's pretty clear what everything is doing. But at this point you can consult docs.gl, if it matters.

Reply:
I got the impression that one of the reasons behind various incarnations of macros which were discussed here at length was these gl calls being hard to follow. The comments are here so that anyone reading this code would be immediately aware of arguments meaning without consulting docs or anything.

If you still feel that they should not be there, I can remove them.

Reply:
> I got the impression that one of the reasons behind various incarnations of macros which were discussed here at length was these gl calls being hard to follow.

This is not the case, instead the reason is to make sure that ordering of functions is guranteed to be correct and work right. Having just a ton of separate functions one after each other makes it harder to grasp which things are connected and need to work together.

> The comments are here so that anyone reading this code would be immediately aware of arguments meaning without consulting docs or anything.

The docs are exactly what this kind of stuff is there for, ideally the function in Rust would make that immediately obvious but unfortunately through FFI that is often lost. But it's not our job here to document OpenGL functions.

> If you still feel that they should not be there, I can remove them.

I think other than "magic numbers" like passing `2`, they shouldn't be there. Extracting them into variables might help with that. But documenting just what function parameters do on the OpenGL side doesn't make much sense.